### PR TITLE
fixing issue 6361

### PR DESCRIPTION
**Pull request by muhdmud**
fixing #6361

### CODE REVIEW

**Comment by muhdmud** [comment]
@ppenzin Hi Petr,

I fixed the oneliner as instructed. Also changed the testcases for toLocaleString.baseline and js in test/Array.
Please advise me if there are other instances i need to change.

Thanks!

**Comment by ppenzin** [comment]
@muhdmud Hi Muhd, preliminarily, the patch does not look too bad, I will take a closer look tomorrow, but we need to sort out the CI failures (see the "Checks" insert).

Note, a couple of the CI jobs have a tendency failing intermittently.

**Comment by muhdmud** [comment]
@ppenzin Hi Petr, apologies i dont get the checks insert part of your statement. Do you mean the tab above between commits and file changed?

**Comment by ppenzin** [comment]
I mean the area in this tab that says "Some checks were not successful", it usually appears after the bottom message in this thread.

Let me know if you need help reproducing the failures.

**Comment by muhdmud** [comment]
Looks like something wrong with my tests, i will check and push again for now

**Comment by muhdmud** [comment]
@ppenzin looks like its failing the tests in toLocaleString.baseline,
What I did was change that oneliner and then changed the tests in above file and remove the spaces.

for example
0.00, anObject, , , another Object, 1.00, a 3rd Object, 2.00 (with spaces)
0.00,anObject,,,another Object,1.00,a 3rd Object,2.00 (removed (spaces)

is this the correct way? also could it be that the function is affecting other tests in other files? so need to change all at once?
Lastly, to run tests before pushing, am i correct to use ./build.sh --test-build?

**Comment by ppenzin** [comment]
To run the tests, run`./runtests.py` in the `test` directory. You would need to add `--test` or `--debug` for test or debug build (in your case it is the test build), and also `--binary <path to ch>`, which the path being something like `out/Test/ch/ch`.

I will take a look at the test tomorrow, I don't think it should be anything serious.

Alternatively you can try to use the experimental end-to-end CMake build for both build and test:

```
$ mkdir <build dir name>
$ cd <build dir name>
$ cmake <path to ChakraCore root>
$ make
$ make check
```

The last command runs the test suite. This would also work with Ninja instead of Make.

**Review by ppenzin — CHANGES_REQUESTED** [review]
Couple of changes probably needed. I think we need the trailing (and only the trailing) spaces back in the .baseline file, to avoid false failures, as output comparison is literal.

However, the reason the tests are failing is that we (myself and @rhuanjl) managed to miss another spot where we are getting the space from. I think after you change the two functions mentioned in https://github.com/microsoft/ChakraCore/issues/6361#issuecomment-698754471 the tests should pass.

File: test/Array/toLocaleString.baseline [hunk 1]
```diff
@@ -1,48 +1,48 @@
 0. Non Object
-1 --> 
-NaN --> 
-true --> 
+1 -->
```
[the remark this answers is not in the document]
Comment (muhdmud) [reply test/Array/toLocaleString.baseline:2]: Oh i see, ok will revert the all the trailing spaces in this file

**Comment by muhdmud** [comment]
for the part where i have to change the function in two places,

if (*sepOutSize < 2 || szSeparator[*sepOutSize - 2] != ' ')
            {
                szSeparator[*sepOutSize - 1] = ' ';    // this line you mean?
                szSeparator[*sepOutSize] = '\0';
            }

**Comment by ppenzin** [comment]
Yeah, you can just delete that whole section, I think.

File: test/Array/toLocaleString.baseline (hunk 1 above)
[the remark this answers is not in the document]
Comment (muhdmud) [reply test/Array/toLocaleString.baseline:2]: readded all the spaces as requested

**Comment by muhdmud** [comment]
I added the trailing spaces for the rest but for when i saved it went away.. especially for

0. Non Object
1 -->(space here but cannot save)
NaN --> 
true -->

**Comment by ppenzin** [comment]
I think it your text editor mandating "no trailing spaces" - there must be a way to turn that off. I've created a gist with the trailing spaces put back, you can try to copy it (but probably don't open with the same editor before `git add` on the file): https://gist.github.com/ppenzin/c375a5edc9e2749c7a751175c81ef123

The code changes look good to me, I think we will be all set once we sort the trailing spaces out.

**Comment by muhdmud** [comment]
Oh ok noted. I did change the trailing spaces settings in VSCode, thanks for that 👍
Now pushed again with the trailing spaces, thanks sir!

**Review by ppenzin — APPROVED** [review]
You are welcome! :+1: I think the change should be good now, but we might need to get through the usual continuous integration issues.

Edit: Windows 7 failure is a bit worrying, something minor might be still off, but at the very least we are getting really close.

**Comment by muhdmud** [comment]
Sure, il do what it takes! quite new so its good learning experience for me

**Review by ppenzin — CHANGES_REQUESTED** [review]
> Edit : looks like I need to readd trailing for the other testcases as well? not only first part

That is true, even though it looks similar, it is a different test, it is in `test/typedarray`.

Sorry, I have also jumped the gun on POSIX array utilities (see comment on the file). I am not sure if a similar change is needed fro the Windows array utilities.

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/ArraysUtility.cpp [hunk 2]
```diff
@@ -53,8 +53,6 @@ namespace Arrays
         ArrayLocalization arrayLocalization;
         //Append ' ' after separator
         szSeparator[*sepOutSize] = arrayLocalization.GetLocaleSeparator();
-        szSeparator[++(*sepOutSize)] = ' ';
-        szSeparator[++(*sepOutSize)] = '\0';
```
[the remark this answers is not in the document]
Comment (muhdmud) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/ArraysUtility.cpp:57]: Sure, il just add the posix one for now

[the remark this answers is not in the document]
Comment (muhdmud) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/ArraysUtility.cpp:57]: As you pointed out the issue with typedarray, it is here
-----------------------------------------------------------------
                // On OSX and Linux the values are printed as 0 instead 0.00. This is a valid workaround as we have still validated the toLocaleString behavior is correct.
                if (result == "0, 31") {
                    result = "0.00, 31.00";
                }

                assert.areEqual("0.00, 31.00", result, "TypedArray" + helpers.getTypeOf(o) + ".toLocaleString should use length from internal slot.");
                return result;

---------------------------------------------------

[the remark this answers is not in the document]
Comment (ppenzin) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/ArraysUtility.cpp:57]: You need to change that line, like you changed the other .js test - that should take care at least of POSIX tests.

[the remark this answers is not in the document]
Comment (muhdmud) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/ArraysUtility.cpp:57]: posix macos looks good now! so for windows do i need to readd the string terminator as well?

**Review by ppenzin — APPROVED** [review]

**Comment by ppenzin** [comment]
I am happy we got past the POSIX issue. The Windows issue is a bit weird.

Have you looked at what is causing Windows test to fail? ;)  If you can, try to print both expected and result character by character or run the test under debugger.

File: lib/Runtime/Library/JavascriptArray.cpp [hunk 3]
```diff
@@ -8390,7 +8390,7 @@ using namespace Js;
                 }
                 else
                 {
-                    separator = scriptContext->GetLibrary()->GetCommaSpaceDisplayString();
```
Comment (ppenzin) [inline lib/Runtime/Library/JavascriptArray.cpp:8393]: Locale separator is produced on this line (above):
```
bool hasLocaleSeparator = Arrays::GetLocaleSeparator(szSeparator, &sepSize, Arrays::SeparatorBufferSize);
```

Comment (ppenzin) [inline lib/Runtime/Library/JavascriptArray.cpp:8393]: Locale separator is produced on this line (above):
```
bool hasLocaleSeparator = Arrays::GetLocaleSeparator(szSeparator, &sepSize, Arrays::SeparatorBufferSize);
```
Also, this line is key to understanding how we use its size:
```
separator = JavascriptString::NewCopyBuffer(szSeparator, static_cast<charcount_t>(sepSize), scriptContext);
```

File: lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp [hunk 4]
```diff
@@ -31,16 +31,6 @@ namespace Arrays
             AssertMsg(FALSE, "GetLocaleInfo failed");
             return false;
         }
-        else
-        {
-            // Append ' ' if necessary
-            if (*sepOutSize < 2 || szSeparator[*sepOutSize - 2] != ' ')
```
Comment (ppenzin) [inline lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp:37]: Hint to unlock the Windows test failure: what value would `*sepOutSize` have when separator is just a string containing a single character and would that work for what it is used for in `JavaScriptArray.cpp`? (see my other comment)  `sepOutSize` comes out of a system function (`GetLocaleInfoEx`) on line 27 above; MSDN has a documentation page on it.

**Review by ppenzin — COMMENTED** [review]
Sorry, this is definitely turning from a trivial couple of lines change into something more challenging. I've left you some hints about how to resolve remaining test failures - let me know if you have questions or they are a bit hard to follow.

Please don't be discouraged if you don't understand how to proceed and ask for help - the goal is for everybody to learn something, rather than getting the right answer :)

File: lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp (hunk 4 above)
Comment (muhdmud) [reply lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp:37]: if i understand correctly, seems like we using the GetLocaleInfoEx to get the size of the locale separator, then we use that size to copy over from scriptcontext to our szSeparator buffer?

from msdn, since cchdata that is passed into the GetLocaleInfoEx is unsigned 32bit int value of 6, so sepOutSize will have value of 2?
looks like need to append that null terminating character to the szSeparator in the else statement, if not it wonr even go into that JavaScriptArray.cpp if scope

Comment (ppenzin) [reply lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp:37]: Yes, you are right - the system call on Windows counts the null character, while the consuming code in JavaScriptArray (and  other places) does not. We just need to make the size here match what the consuming code expects. Right now it seem to copy null character in JavaScriptArray. It might be a good idea to add a comment that we expect to not count null character.

Comment (muhdmud) [reply lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp:37]: Oh I see, thanks! So can I just add the else statement again and add the null character when sepOutSize is 1 now like below?

else{
            if (*sepOutSize < 2)
            {
                szSeparator[*sepOutSize] = '\0';
            }
}

Comment (ppenzin) [reply lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp:37]: I think you need to decrement `*sepOutSize`, regardless of what the size is, since syscall returns value too high by one:

```
+        else
+        {
+            // We need just the number of regular characters, but Win32 API counts terminating null character as well
+            --(*sepOutSize);
+        }
+
```

Comment (muhdmud) [reply lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp:37]: Thank for the help! 😊 Where to go from here? do I close this? or wait for it to get merged

Comment (ppenzin) [reply lib/Runtime/PlatformAgnostic/Platform/Windows/ArraysUtility.cpp:37]: It would get closed when the change gets merged, and then I'll check if the issue gets closed too :)

**Review by ppenzin — APPROVED** [review]

**Comment by ppenzin** [comment]
@muhdmud thank you for your contribution!
